### PR TITLE
feat(libc): windows native module support

### PR DESCRIPTION
**Pull request by zeromake**
by  #486

add definde `QJS_NATIVE_MODULE_SUFFIX`, windows is `.dll`, !windows is `.so`.

- [x] add cmake test

### CODE REVIEW

**Comment by zeromake** [comment]
mingw it seems to be different from MSVC……

**Comment by saghul** [comment]
Yep, some if-defing might be required...

**Comment by zeromake** [comment]
@saghul ci is ok

**Comment by saghul** [comment]
Can you make sure the relevant tests also run on Windows?

IIRC they are not compiled in.

**Comment by zeromake** [comment]
@saghul 
That's what you're talking about? https://github.com/quickjs-ng/quickjs/blob/0e5e9c2c49db15ab9579edeb4d90e610c8b8463f/CMakeLists.txt#L295

File: quickjs-libc.c [hunk 1]
```diff
@@ -479,7 +488,54 @@ typedef JSModuleDef *(JSInitModuleFunc)(JSContext *ctx,
...
-#if defined(_WIN32) || defined(__wasi__)
+#if defined(_WIN32)
+static JSModuleDef *js_module_loader_so(JSContext *ctx,
+                                        const char *module_name)
+{
+    JSModuleDef *m;
+    HINSTANCE hd;
+    JSInitModuleFunc *init;
+    char *filename = NULL;
+    int len = strlen(module_name);
```
Comment (bnoordhuis) [inline quickjs-libc.c]:
```suggestion
    size_t len = strlen(module_name);
```

File: quickjs-libc.c [hunk 2]
```diff
@@ -479,7 +488,54 @@ typedef JSModuleDef *(JSInitModuleFunc)(JSContext *ctx,
...
+    if (len > 2 &&
+        ((module_name[0] >= 'A' && module_name[0] <= 'Z') ||
+            (module_name[0] >= 'a' && module_name[0] <= 'z')) &&
+        module_name[1] == ':') {
+        filename = (char *)module_name;
+    } else if (len > 2 && module_name[0] != '.' && (module_name[1]!= '/' || module_name[1] == '\\')) {
```
Comment (bnoordhuis) [inline quickjs-libc.c]: This doesn't look correct to me.  What exactly are you trying to accept/reject here?

File: quickjs-libc.c [hunk 3]
```diff
@@ -479,7 +488,54 @@ typedef JSModuleDef *(JSInitModuleFunc)(JSContext *ctx,
...
+        filename = js_malloc(ctx, strlen(module_name) + 2 + 1);
+        if (!filename)
+            return NULL;
+        strcpy(filename, "./");
+        strcpy(filename + 2, module_name);
+    }
+    {
+        hd = LoadLibraryA(filename);
```
Comment (bnoordhuis) [inline quickjs-libc.c]: filename can be NULL because of the if/else if block above. Why is it inside a scope?

File: quickjs-libc.c [hunk 4]
```diff
@@ -479,7 +488,54 @@ typedef JSModuleDef *(JSInitModuleFunc)(JSContext *ctx,
...
+        if (filename!= module_name)
```
Comment (bnoordhuis) [inline quickjs-libc.c]:
```suggestion
        if (filename != module_name)
```

File: quickjs-libc.c [hunk 5]
```diff
@@ -479,7 +488,54 @@ typedef JSModuleDef *(JSInitModuleFunc)(JSContext *ctx,
...
+            js_free(ctx, filename);
+        if (hd == NULL) {
+            JS_ThrowReferenceError(ctx, "js_load_module '%s' error: %lu",
+                                module_name, GetLastError());
```
Comment (bnoordhuis) [inline quickjs-libc.c]:
```suggestion
                                    module_name, GetLastError());
```
I believe that lines up the arguments but please check.

File: quickjs-libc.c [hunk 6]
```diff
@@ -479,7 +488,54 @@ typedef JSModuleDef *(JSInitModuleFunc)(JSContext *ctx,
...
+            goto fail;
+        }
+    }
+    init = (JSInitModuleFunc *)(void *)GetProcAddress(hd, "js_init_module");
+    if (!init) {
+        JS_ThrowReferenceError(ctx, "js_init_module '%s' not found: %lu",
+                               module_name, GetLastError());
```
Comment (bnoordhuis) [inline quickjs-libc.c:525]: It'd be nicer to use FormatMessageA() to get a human-readable error message ([example](https://github.com/libuv/libuv/blob/58dfb6c89ba3deb23301ddbdd4a8769f4c767dff/src/win/dl.c#L105-L109))

**Comment by saghul** [comment]
> @saghul 
> 
> That's what you're talking about? https://github.com/quickjs-ng/quickjs/blob/0e5e9c2c49db15ab9579edeb4d90e610c8b8463f/CMakeLists.txt#L295 

Yep those will do!

File: quickjs-libc.c (hunk 2 above)
Comment (zeromake) [reply quickjs-libc.c]: @bnoordhuis 
It's me who wrote it wrong, and it's not needed in either case `./`

File: quickjs-libc.c (hunk 6 above)
Comment (zeromake) [reply quickjs-libc.c:525]: @bnoordhuis 
FormatMessageA error message fetching is a bit complicated, whether to consider a separate method to get it
```c
struct JSContext {
    char* errmsg;
}
static int js_error(JSContext *ctx) {
    if (ctx->errmsg) {
    LocalFree(ctx->errmsg);
    ctx->errmsg = NULL;
  }
  // FormatMessageA ……
}
```

**Comment by zeromake** [comment]
https://github.com/quickjs-ng/quickjs/actions/runs/10589030031/job/29348830278 ……

**Comment by saghul** [comment]
How weird, I'll take a look at that one. You can ignore it.

**Comment by saghul** [comment]
I guess what's missing is to run the actual tests, check this one: https://github.com/quickjs-ng/quickjs/blob/0e5e9c2c49db15ab9579edeb4d90e610c8b8463f/.github/workflows/ci.yml#L137

File: quickjs-libc.c [hunk 7]
```diff
@@ -479,7 +488,52 @@ typedef JSModuleDef *(JSInitModuleFunc)(JSContext *ctx,
...
+                                        const char *module_name)
+{
+    JSModuleDef *m;
+    HINSTANCE hd;
+    JSInitModuleFunc *init;
+    char *filename = NULL;
+    size_t len = strlen(module_name);
+    JS_BOOL is_absolute = len > 2 && ((module_name[0] >= 'A' && module_name[0] <= 'Z') ||
+            (module_name[0] >= 'a' && module_name[0] <= 'z')) && module_name[1] == ':';
+    JS_BOOL is_relative = len > 2 && module_name[0] != '.' && (module_name[1] == '/' || module_name[1] == '\\');
```
Comment (bnoordhuis) [inline quickjs-libc.c]: Shouldn't that be:
```suggestion
    JS_BOOL is_relative = len > 2 && module_name[0] == '.' && (module_name[1] == '/' || module_name[1] == '\\');
```

File: quickjs-libc.c [hunk 8]
```diff
@@ -479,7 +488,52 @@ typedef JSModuleDef *(JSInitModuleFunc)(JSContext *ctx,
...
+    }
+    hd = LoadLibraryA(filename);
+    if (filename != module_name)
+        js_free(ctx, filename);
+    if (hd == NULL) {
+        JS_ThrowReferenceError(ctx, "js_load_module '%s' error: %lu",
+                               module_name, GetLastError());
+        goto fail;
+    }
+    init = (JSInitModuleFunc *)(void *)GetProcAddress(hd, "js_init_module");
```
Comment (bnoordhuis) [inline quickjs-libc.c]:
```suggestion
    init = (JSInitModuleFunc *)(uintptr_t)GetProcAddress(hd, "js_init_module");
```
Otherwise mingw will probably complain in pedantic mode.

File: quickjs-libc.c [hunk 9]
```diff
@@ -479,7 +488,52 @@ typedef JSModuleDef *(JSInitModuleFunc)(JSContext *ctx,
...
+        JS_ThrowReferenceError(ctx, "js_init_module '%s' not found: %lu",
+                               module_name, GetLastError());
+        goto fail;
+    }
+    m = init(ctx, module_name);
+    if (!m) {
+        JS_ThrowReferenceError(ctx, "js_call_module '%s' initialization error",
+                               module_name);
+    fail:
+        if (hd != NULL) FreeLibrary(hd);
```
Comment (bnoordhuis) [inline quickjs-libc.c]:
```suggestion
        if (hd != NULL)
            FreeLibrary(hd);
```

**Comment by zeromake** [comment]
@saghul 
[job](https://github.com/quickjs-ng/quickjs/actions/runs/10692515096/job/29649978234) error:
```sh
/usr/bin/ld: libqjs.a(quickjs.c.o): relocation R_X86_64_PC32 against symbol `digits36' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: bad value
```
linux build so file need `-fPIC` flags?

**Comment by saghul** [comment]
Not sure why that started happening, we need to take a look, but you can ignore that for this PR.

File: CMakeLists.txt [hunk 10]
```diff
@@ -292,27 +292,27 @@ if(BUILD_EXAMPLES AND NOT WIN32)
...
-        if(APPLE)
-            target_link_options(fib PRIVATE -undefined dynamic_lookup)
-        endif()
+    add_library(fib MODULE examples/fib.c)
+    set_target_properties(fib PROPERTIES
+        PREFIX ""
+        C_VISIBILITY_PRESET default
+    )
+    target_compile_definitions(fib PRIVATE JS_SHARED_LIBRARY)
+    target_link_libraries(fib ${qjs_libs})
```
Comment (saghul) [inline CMakeLists.txt]: Ah this is what causes the linking error in Ubuntu, actually. On unix we don't need to link the module with the library.

Comment (zeromake) [reply CMakeLists.txt]: Maybe it should be?
``` diff
-target_link_libraries(fib ${qjs_libs})
+if(WIN32)
+      target_link_libraries(fib ${qjs_libs})
+endif()
```

Comment (saghul) [reply CMakeLists.txt]: I think so yeah

Comment (zeromake) [reply CMakeLists.txt]: @saghul 
[change link](https://github.com/quickjs-ng/quickjs/pull/490/commits/096ada0a027da304788b2c1f4756bfbf2948e538)

**Comment by zeromake** [comment]
@saghul 
https://github.com/quickjs-ng/quickjs/pull/490/commits/c2391b091de558fbbb503f7338d952d9f6152045
- windows shared use `.so`
- `fib.c` use` __declspec(dllexport)` js_init_module
- ci add examples\test_fib.js

**Comment by zeromake** [comment]
@saghul 
except for the wrong text, It should all be done, and you can consider whether to merge or not.

File: CMakeLists.txt [hunk 11]
```diff
@@ -292,27 +292,33 @@ if(BUILD_EXAMPLES AND NOT WIN32)
...
-        endif()
+    add_library(fib MODULE examples/fib.c)
+    set_target_properties(fib PROPERTIES
+        PREFIX ""
+        C_VISIBILITY_PRESET default
+    )
+    target_compile_definitions(fib PRIVATE JS_SHARED_LIBRARY)
+    if(WIN32)
+        target_link_libraries(fib ${qjs_libs})
+        set_target_properties(fib PROPERTIES SUFFIX ".so")
```
Comment (saghul) [inline CMakeLists.txt]: Why not leave the default (DLL) ?

File: quickjs-libc.c [hunk 12]
```diff
@@ -94,6 +94,10 @@ extern char **environ;
 #include "list.h"
 #include "quickjs-libc.h"
 
+#ifndef QJS_NATIVE_MODULE_SUFFIX
+#define QJS_NATIVE_MODULE_SUFFIX ".so"
```
Comment (saghul) [inline quickjs-libc.c:103]: Why did you go back on having different suffixes on windows? I think it makes sense to use the platform default.

Comment (zeromake) [reply quickjs-libc.c:103]: @saghul 
[test_point.js](https://github.com/quickjs-ng/quickjs/blob/master/examples/test_point.js) use esm `import`
but [esm](https://nodejs.org/api/esm.html#import-specifiers) not support dynamic splicing.
```js
import * as os from "os";
const isWin = os.platform === 'win32';
import { fib } from `./fib.${isWin ? 'dll' : 'so'}`;

console.log("Hello World");
console.log("fib(10)=", fib(10));

```
``` sh
SyntaxError: string expected
    at ./examples/test_fib.js:4:1
```

Comment (saghul) [reply quickjs-libc.c:103]: You can use a dynamic import and await it.

Comment (zeromake) [reply quickjs-libc.c:103]: @saghul ok, this is [change](https://github.com/quickjs-ng/quickjs/pull/490/commits/fedc1bc8ede6d18f2f78c394c66eee6564745f3a)

**Comment by zeromake** [comment]
@saghul [job/29858936901](https://github.com/quickjs-ng/quickjs/actions/runs/10768599200/job/29858936901) What is wrong with this

**Comment by saghul** [comment]
Since you modified the JS part of the test, you need to commit the result of `make codegen`.

**Comment by chqrlie** [comment]
> Since you modified the JS part of the test, you need to commit the result of `make codegen`.

Yes, this is unfortunate.  We should try and find a way to generate these files at build time and avoid committing them.  This would also reduce bloat in the repository.

**Comment by saghul** [comment]
> Yes, this is unfortunate. We should try and find a way to generate these files at build time and avoid committing them. This would also reduce bloat in the repository.

I have the same conundrum on txiki.js, only worse, because I have a lot more generated code.

I gave this a long hard thought and while we could generate it in the release job and then make a "release tarball", I'd rather have the git repo be the source of truth, specially after the xz fiasco.

**Comment by bnoordhuis** [comment]
I concur with that sentiment, fwiw.

**Comment by saghul** [comment]
> I concur with that sentiment, fwiw.

Which of the 2? 😅

**Comment by chqrlie** [comment]
> I gave this a long hard thought and while we could generate it in the release job and then make a "release tarball", I'd rather have the git repo be the source of truth, specially after the xz fiasco.

I am not sure I understand the connection with the xz fiasco.  We shall be very cautious about accepting PRs from, let alone giving commit rights to outsiders.  

Generating the byte code files at build time is mainly a problem for cross compiling where it requires building a qjs executable for the host in addition to cross compiling for the selected target.  This used to be problematic because the generated files depended on the endianness of the target, which is no longer the case.

**Comment by bnoordhuis** [comment]
> Which of the 2? 😅

That git should be the Source Of Truth :)

(I deliberated on whether to mention I was referring to your comment but, since a couple of a minutes had passed, I figured it was unambiguously in response to the last comment.)

**Comment by saghul** [comment]
Let's continue here: https://github.com/quickjs-ng/quickjs/issues/509 not to derail the PR.

**Comment by saghul** [comment]
@zeromake There seems to be a conflict. Can you please rebase with latest master?

**Comment by zeromake** [comment]
@saghul is merged

File: .github/workflows/ci.yml [hunk 13]
```diff
@@ -192,14 +192,20 @@ jobs:
       - uses: actions/checkout@v4
       - name: build
         run: |
-          cmake -B build -G "Visual Studio 17 2022" -A ${{matrix.arch}}
+          cmake -B build -D BUILD_EXAMPLES=ON -D BUILD_QJS_LIBC=ON -G "Visual Studio 17 2022" -A ${{matrix.arch}}
```
Comment (saghul) [inline .github/workflows/ci.yml]: I don't think you need to specify `BUILD_QJS_LIBC=ON`

File: quickjs-libc.c [hunk 14]
```diff
@@ -479,7 +489,53 @@ typedef JSModuleDef *(JSInitModuleFunc)(JSContext *ctx,
...
+        filename = js_malloc(ctx, len + 2 + 1);
+        if (!filename)
+            return NULL;
+        strcpy(filename, "./");
+        strcpy(filename + 2, module_name);
+    }
+    hd = LoadLibraryA(filename);
+    if (filename != module_name)
+        js_free(ctx, filename);
+    if (hd == NULL) {
```
Comment (saghul) [inline quickjs-libc.c:517]: Maybe do this before freeing filename, and log filename instead?

File: quickjs-libc.c (hunk 6 above)
Comment (saghul) [reply quickjs-libc.c:525]: @bnoordhuis Doing it The Right Way (TM) seems a bit involved: https://github.com/libuv/libuv/blob/0a00e80c3686b93eccb9a801954e86bd7d7fe6ab/src/win/dl.c#L92 perhaps we can simplify here?

File: quickjs-libc.c (hunk 14 above)
Comment (zeromake) [reply quickjs-libc.c:517]: @saghul 
> Maybe do this before freeing filename, and log filename instead?

I am the reference unix js_module_loader_so
```c
    hd = dlopen(filename, RTLD_NOW | RTLD_LOCAL);
    if (filename != module_name)
        js_free(ctx, filename);
    if (!hd) {
        JS_ThrowReferenceError(ctx, "could not load module filename '%s' as shared library: %s",
                               module_name, dlerror());
        goto fail;
    }
```

Comment (saghul) [reply quickjs-libc.c:517]: ok

**Review by saghul — APPROVED** [review]
Great work! @bnoordhuis can you PTAL too?

File: .github/workflows/ci.yml [hunk 15]
```diff
@@ -192,14 +192,20 @@ jobs:
       - uses: actions/checkout@v4
       - name: build
         run: |
-          cmake -B build -G "Visual Studio 17 2022" -A ${{matrix.arch}}
+          cmake -B build -D BUILD_EXAMPLES=ON -G "Visual Studio 17 2022" -A ${{matrix.arch}}
```
Comment (bnoordhuis) [inline .github/workflows/ci.yml]: For consistency, write as:
```suggestion
          cmake -B build -DBUILD_EXAMPLES=ON -G "Visual Studio 17 2022" -A ${{matrix.arch}}
```

File: .github/workflows/ci.yml [hunk 16]
```diff
@@ -221,14 +227,20 @@ jobs:
       - uses: actions/checkout@v4
       - name: build
         run: |
-          cmake -B build -G "Visual Studio 17 2022" -T ClangCL
+          cmake -B build -D BUILD_EXAMPLES=ON -G "Visual Studio 17 2022" -T ClangCL
```
Comment (bnoordhuis) [inline .github/workflows/ci.yml]: Ditto
```suggestion
          cmake -B build -DBUILD_EXAMPLES=ON -G "Visual Studio 17 2022" -T ClangCL
```

File: .github/workflows/ci.yml [hunk 17]
```diff
@@ -254,14 +266,20 @@ jobs:
           ninja.exe --version
       - name: build
         run: |
-          cmake -B build -DCMAKE_BUILD_TYPE=${{matrix.buildType}} -G "Ninja"
+          cmake -B build -D BUILD_EXAMPLES=ON -DCMAKE_BUILD_TYPE=${{matrix.buildType}} -G "Ninja"
```
Comment (bnoordhuis) [inline .github/workflows/ci.yml]: And ditto
```suggestion
          cmake -B build -DBUILD_EXAMPLES=ON -DCMAKE_BUILD_TYPE=${{matrix.buildType}} -G "Ninja"
```

File: CMakeLists.txt [hunk 18]
```diff
@@ -287,27 +287,31 @@ if(BUILD_EXAMPLES AND NOT WIN32)
...
-            target_link_options(fib PRIVATE -undefined dynamic_lookup)
-        endif()
+    add_library(fib MODULE examples/fib.c)
+    set_target_properties(fib PROPERTIES
+        PREFIX ""
+        C_VISIBILITY_PRESET default
+    )
+    target_compile_definitions(fib PRIVATE JS_SHARED_LIBRARY)
+    if(WIN32)
+        target_link_libraries(fib ${qjs_libs})
```
Comment (bnoordhuis) [inline CMakeLists.txt:300]: For my curiosity, what library or libraries in particular are necessary? You normally only use qjs_libs for binaries, not shared libraries.

File: examples/fib.c [hunk 19]
```diff
@@ -61,7 +61,15 @@ static int js_fib_init(JSContext *ctx, JSModuleDef *m)
...
 #endif
 
-JSModuleDef *JS_INIT_MODULE(JSContext *ctx, const char *module_name)
+#ifndef JS_EXTERN
+#ifdef _WIN32
+#define JS_EXTERN __declspec(dllexport)
+#else
+#define JS_EXTERN
+#endif
+#endif
```
Comment (bnoordhuis) [inline examples/fib.c:70]: JS_EXTERN is always defined, right? Why is this necessary?

File: quickjs-libc.c [hunk 20]
```diff
@@ -94,6 +94,16 @@ extern char **environ;
...
 
+#ifdef _WIN32
+#ifndef QJS_NATIVE_MODULE_SUFFIX
+#define QJS_NATIVE_MODULE_SUFFIX ".dll"
+#endif
+#else
+#ifndef QJS_NATIVE_MODULE_SUFFIX
+#define QJS_NATIVE_MODULE_SUFFIX ".so"
+#endif
+#endif
```
Comment (bnoordhuis) [inline quickjs-libc.c]: Why define this conditionally? And why not more succinctly like this?
```suggestion
#ifndef QJS_NATIVE_MODULE_SUFFIX
#ifdef _WIN32
#define QJS_NATIVE_MODULE_SUFFIX ".dll"
#else
#define QJS_NATIVE_MODULE_SUFFIX ".so"
#endif
#endif
```

**Review by bnoordhuis — APPROVED** [review]
LGTM with some suggestions and questions.

File: CMakeLists.txt (hunk 18 above)
Comment (saghul) [reply CMakeLists.txt:300]: I looked at this at one point. The problem seems to be weak symbol linking is not a thing on windows (?) or I didn't know the right flags!

Comment (bnoordhuis) [reply CMakeLists.txt:300]: Is there something (like functions from math.h) leaking out of quickjs.h that necessitates that? Or is it something else?

Comment (saghul) [reply CMakeLists.txt:300]: IIrC it's the quickjs functions themselves.

Comment (bnoordhuis) [reply CMakeLists.txt:300]: Okay, that makes sense, but wouldn't it be best to link against just quickjs.lib?

Comment (saghul) [reply CMakeLists.txt:300]: Fair point!

File: examples/fib.c (hunk 19 above)
Comment (saghul) [reply examples/fib.c:70]: Indeed, let's export this properly in quickjs.h @zeromake

Comment (zeromake) [reply examples/fib.c:70]: @saghul 
I tried `#include "../quickjs.h"` use `JS_EXTERN`
``` sh
/Users/zero/project/quickjs-ng/examples/fib.c:72:1: error: unknown type name 'JS_EXTERN'
JS_EXTERN JSModuleDef *JS_INIT_MODULE(JSContext *ctx, const char *module_name)
^
/Users/zero/project/quickjs-ng/examples/fib.c:72:22: error: expected ';' after top level declarator
JS_EXTERN JSModuleDef *JS_INIT_MODULE(JSContext *ctx, const char *module_name)
```
on quickjs.h has `JS_EXTERN` ……

Comment (zeromake) [reply examples/fib.c:70]: quickjs.h latest is `#undef JS_EXTERN` ……
https://github.com/quickjs-ng/quickjs/blob/5bf35450cd0e03b9a6576f3d9db22b1e1ea20548/quickjs.h#L1026

Comment (zeromake) [reply examples/fib.c:70]: @saghul @bnoordhuis 
https://github.com/quickjs-ng/quickjs/pull/490/commits/a9239349a095ac3b074ab0cf2ba5f897648f1f25
del `#undef JS_EXTERN`

**Comment by saghul** [comment]
```
LINK : fatal error LNK1149: output filename matches input filename 'D:\a\quickjs\quickjs\build\Release\qjs.lib' [D:\a\quickjs\quickjs\build\qjs_exe.vcxproj]
```

WTF?!

**Comment by zeromake** [comment]
> ```
> LINK : fatal error LNK1149: output filename matches input filename 'D:\a\quickjs\quickjs\build\Release\qjs.lib' [D:\a\quickjs\quickjs\build\qjs_exe.vcxproj]
> ```
> 
> WTF?!

i do rollback
```diff
+target_link_libraries(fib qjs)
-target_link_libraries(fib ${qjs_libs})
```

**Comment by zeromake** [comment]
@saghul [runs/10788294143](https://github.com/quickjs-ng/quickjs/actions/runs/10788294143)  run on 3 hours ago……

**Comment by saghul** [comment]
I think our free credit is running low 😅

**Comment by saghul** [comment]
Hum I see the failure again. I think it might be the JS_EXTERN change. Can you undo that and let's check again?

**Comment by saghul** [comment]
Looking at what we do in libuv:

```
#ifndef UV_EXTERN
#ifdef _WIN32
  /* Windows - set up dll import/export decorators. */
# if defined(BUILDING_UV_SHARED)
    /* Building shared library. */
#   define UV_EXTERN __declspec(dllexport)
# elif defined(USING_UV_SHARED)
    /* Using shared library. */
#   define UV_EXTERN __declspec(dllimport)
# else
    /* Building static library. */
#   define UV_EXTERN /* nothing */
# endif
#elif __GNUC__ >= 4
# define UV_EXTERN __attribute__((visibility("default")))
#elif defined(__SUNPRO_C) && (__SUNPRO_C >= 0x550) /* Sun Studio >= 8 */
# define UV_EXTERN __global
#else
# define UV_EXTERN /* nothing */
#endif
#endif /* UV_EXTERN */
```

I think we should undo the JS_EXTERN change and likely do something like the above later on.

**Comment by saghul** [comment]
@bnoordhuis I'll make some follow-up refinements since I'm playing around with the linking libraries in another PR.

**Comment by saghul** [comment]
Nice work @zeromake ! 👏
